_tutorials/spying_on_properties.md

```diff
@@ -0,0 +1,59 @@
+---
+layout: article
+title: Spying on properties
+---
+
+# Spying on properties
+
+Properties are more complicated than functions. In Jasmine, you can do anything with a property spy that you
+can do with a function spy, but you may need to use different syntax.
+
+Use `spyOnProperty` to create either a getter or setter spy.
+
+```javascript
+it("allows you to create spies for either type", function() {
+  spyOnProperty(someObject, "myValue", "get").and.returnValue(30);
+  spyOnProperty(someObject, "myValue", "set").and.callThrough();
+});
+```
+
+Changing the value of an existing spy is more difficult than with a function, because you cannot _refer_ to a property without calling its `getter` method. To get around this, you can save a referene to the spy for later changes.
+
+```javascript
+beforeEach(function() {
+  this.propertySpy = spyOnProperty(someObject, "myValue", "get").and.returnValue(1);
+});
+
+it("lets you change the spy strategy later", function() {
+  this.propertySpy.and.returnValue(3);
+  expect(someObject.myValue).toEqual(3);
+});
+```
+
+If saving a reference to the spy is awkward, you can also access it from anywhere in your test using
+`Object.getOwnPropertyDescriptor`.
+
+```javascript
+beforeEach(function() {
+  spyOnProperty(someObject, "myValue", "get").and.returnValue(1);
+});
+
+it("lets you change the spy strategy later", function() {
+  Object.getOwnPropertyDescriptor(someObject, "myValue").get.and.returnValue(3);
+  expect(someObject.myValue).toEqual(3);
+});
+```
+
+You can create a spy object with several properties on it quickly by passing an array or hash of properties as
+a third argument to `createSpyObj`. In this case you won't have a reference to the created spies, so if you
+need to change their spy strategies later, you can use the `Object.getOwnPropertyDescriptor` approach.
+
+```javascript
+it("creates a spy object with properties", function() {
+  let obj = createSpyObj("myObject", {}, { x: 3, y: 4 });
+  expect(obj.x).toEqual(3);
+
+  Object.getOwnPropertyDescriptor(obj, "x", "get").and.returnValue(7);
```

Comment:
`Object.getOwnPropertyDescriptor` has only two params.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor

When i tried this i got en error:

> Cannot read property 'returnValue' of undefined

I think it should be:
```
    Object.getOwnPropertyDescriptor(obj, "x").get.and.returnValue(7);
```

Reply:
🤦 Thanks for catching! Will put up a correction.

Reply:
Object.getOwnPropertyDescriptor(obj, "x").get.and.returnValue(7); gives error as `Property 'and' does not exist on type '() => any'.`

Reply:
@souptikmakarov it's a typescript issue, quick "fix" to add `// @ts-ignore` before the row. Like this:
```
      const obj = createSpyObj('myObject', {}, { x: 3, y: 4 });
      expect(obj.x).toEqual(3);
      // @ts-ignore
      Object.getOwnPropertyDescriptor(obj, 'x').get.and.returnValue(7);
      expect(obj.x).toEqual(7);
```
Or just cast to any (i think it's uglier):
```
      (Object.getOwnPropertyDescriptor(obj, 'x').get as any).and.returnValue(7);
```